### PR TITLE
Feature/improve num trees so it doesnt switch to force

### PR DESCRIPTION
**Pull request by ThijsLuttikhuis**
- [x] The tests are passing

Refactor Numtrees
Remove Force
Improve Numtrees
???
Profit

Needs #494

### CODE REVIEW

File: roboteam_ai/src/control/numTrees/Collision.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "Collision.h"
+
+namespace rtt {
+namespace ai {
+namespace control {
+
+std::string Collision::collisionTypeToString() {
+    std::string s;
+    switch (getCollisionType()) {
```
Comment (Barisimre) [inline roboteam_ai/src/control/numTrees/Collision.cpp]: Make default

File: roboteam_ai/src/control/numTrees/NumTreePosControl.h [hunk 2]
```diff
@@ -28,16 +26,28 @@ class NumTreePosControl : public ForcePosControl {
...
+        bool checkEmptyPath();
+        bool checkIfTargetMoved(double maxTargetDeviation, const Vector2 &targetPos);
+        bool checkIfAtEndOfPath(double maxTargetDeviation, const Vector2 &targetPos);
+        bool checkIfTooFarFromCurrentPath(double maxTargetDeviation, const Vector2 &vector2);
+
+        RobotCommand computeCommand(const Vector2 &exactTargetPos);
 
         // constants
-        const double MAX_CALCULATION_TIME = 25.0;         // Max calculation time in ms
+        const double MAX_CALCULATION_TIME = 20.0;         // Max calculation time in ms
```
Comment (Barisimre) [inline roboteam_ai/src/control/numTrees/NumTreePosControl.h]: Make this in Constants.cpp

File: roboteam_ai/src/control/numTrees/NumTreePosControl.h [hunk 3]
```diff
@@ -28,16 +26,28 @@ class NumTreePosControl : public ForcePosControl {
...
         double DT = 0.1;                          // timestep for ODE model
         static constexpr double DEFAULT_ROBOT_COLLISION_RADIUS = 0.25; // 3x robot radius
```
Comment (Barisimre) [inline roboteam_ai/src/control/numTrees/NumTreePosControl.h]: Also make this Constant dependent

File: roboteam_ai/src/control/positionControllers/RobotCommand.h [hunk 4]
```diff
@@ -7,17 +7,16 @@
...
 class RobotCommand {
     public:
-        int id = -1;
+        int id = - 1;
+        Vector2 pos = Vector2();
         Vector2 vel = Vector2();
         Angle angle = Angle();
-
         unsigned char dribbler = 0;
         unsigned char geneva = 3;
```
Comment (Barisimre) [inline roboteam_ai/src/control/positionControllers/RobotCommand.h]: Maybe make something like Constants::neutralGeneva

File: roboteam_ai/src/control/shotControllers/ShotController.cpp [hunk 5]
```diff
@@ -110,55 +108,53 @@ Vector2 ShotController::getPlaceBehindBall(world::Robot robot, Vector2 shotTarge
...
 /// Now we should have the ball and kick it.
-ShotData ShotController::shoot(world::Robot robot, std::pair<Vector2, Vector2> driveLine, Vector2 shotTarget, bool chip,
+RobotCommand ShotController::shoot(world::Robot robot, std::pair<Vector2, Vector2> driveLine, Vector2 shotTarget, bool chip,
         BallSpeed desiredBallSpeed) {
 
     auto ball = world::world->getBall();
 
     // move towards the ball
-    control::PosVelAngle pva = robot.getBasicPosControl()->getPosVelAngle(std::make_shared<world::Robot>(robot),
+    auto pva = robot.getBasicPosControl()->getPosVelAngle(std::make_shared<world::Robot>(robot),
```
Comment (Barisimre) [inline roboteam_ai/src/control/shotControllers/ShotController.cpp]: You should maybe refactor all the pva named variables something else

File: roboteam_ai/src/skills/Attack.cpp [hunk 6]
```diff
@@ -22,16 +22,18 @@ Attack::Attack(string name, bt::Blackboard::Ptr blackboard)
 
 /// Get an update on the skill
 bt::Node::Status Attack::onUpdate() {
-    if (!robot) return Status::Running;
+    if (! robot) return Status::Running;
```
Comment (Barisimre) [inline roboteam_ai/src/skills/Attack.cpp]: meh

File: roboteam_ai/src/skills/Skill.h [hunk 7]
```diff
@@ -6,15 +6,15 @@
...
+#include "roboteam_ai/src/control/positionControllers/RobotCommand.h"
 
 namespace rtt {
 namespace ai {
 
 // forward declare control Utils
 namespace control {
 class ControlUtils;
-struct PosVelAngle;
+struct RobotComman;
```
Comment (Barisimre) [inline roboteam_ai/src/skills/Skill.h]: WHAT THE FLYING HECK

File: roboteam_ai/src/control/numTrees/Collision.cpp (hunk 1 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/numTrees/Collision.cpp]: fixed

File: roboteam_ai/src/control/numTrees/NumTreePosControl.h (hunk 2 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/numTrees/NumTreePosControl.h]: I disagree.
these constants are really specific to numtrees, so there is no reason to do it in constants

File: roboteam_ai/src/control/shotControllers/ShotController.cpp (hunk 5 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/shotControllers/ShotController.cpp]: 123 fix some

File: roboteam_ai/src/world/Field.cpp [hunk 8]
```diff
@@ -94,84 +96,87 @@ std::vector<std::pair<Vector2, Vector2>> Field::getBlockadesMappedToGoal(bool ou
...
             }
-            if (validObstacle ) {
+            if (validObstacle) {
                 // constrain the blockades to within the goal
                 if (point1.y > point2.y) { // point1 is largest
                     point1.y = std::min(point1.y, upperGoalSide.y);
                     point2.y = std::max(point2.y, lowerGoalSide.y);
-                    blockades.emplace_back(std::make_pair(point2,point1)); // the first element in the pair is the smallest
-                } else { // point2 is largest
+                    blockades.emplace_back(
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Field.cpp]: yeah just make this on one line and put the comment somewhere else if you really care about ctrl alt L that much.

File: roboteam_ai/src/world/Ball.h [hunk 9]
```diff
@@ -15,11 +15,13 @@ namespace world {
 
 class WorldData;
 class Robot;
-class Ball {
+class Ball : public std::enable_shared_from_this<Ball> {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.h]: brr I really dont like the enable_shared_from_this in general. Avoid this at all costs.

File: roboteam_ai/src/world/Ball.cpp [hunk 10]
```diff
@@ -163,6 +163,10 @@ void Ball::updateBallPosition(const Ball &oldBall, const WorldData &worldData) {
     }
 }
 
+const Ball::BallPtr Ball::copy() const {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.cpp]: Again, we should really not need a copy function like this.

File: roboteam_ai/src/skills/gotopos/GoToPos.cpp [hunk 11]
```diff
@@ -97,5 +93,8 @@ void GoToPos::onTerminate(Status s) {
     gtpTerminate(s);
 }
 
+GoToPos::~GoToPos() {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/gotopos/GoToPos.cpp]: empty destructor --> Just let the compiler handle it for you?

File: roboteam_ai/src/skills/ballPlacement/BallPlacementReceive.cpp [hunk 12]
```diff
@@ -54,9 +54,9 @@ bt::Node::Status BallPlacementReceive::onUpdate() {
 
 
 void BallPlacementReceive::moveToCatchPosition(Vector2 position) {
-    control::PosVelAngle pva = robot->getNumtreePosControl()->getPosVelAngle(robot, position);
-    command.x_vel = static_cast<float>(pva.vel.x);
-    command.y_vel = static_cast<float>(pva.vel.y);
+    auto robotCommand = robot->getNumtreePosControl()->getPosVelAngle(robot, position);
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/ballPlacement/BallPlacementReceive.cpp]: You know the type, any reason to change it to Auto?

File: roboteam_ai/src/skills/ballPlacement/BallPlacementPass.cpp [hunk 13]
```diff
@@ -57,12 +57,13 @@ bt::Node::Status BallPlacementPass::onUpdate() {
...
         if (coach::g_pass.isReadyToReceivePass()) {
-            robot->getShotController()->makeCommand(robot->getShotController()->getShotData(*robot, getKicker(), false), command);
+            auto shotData = robot->getShotController()->getShotData(*robot, getKicker(), false);
+            command = shotData.makeROSCommand();
         } else if (robot->pos.dist(ball->pos) > 0.5) {
-            auto pva = robot->getNumtreePosControl()->getPosVelAngle(robot, ball->pos);
-            command.x_vel = pva.vel.x;
-            command.y_vel = pva.vel.y;
-            command.w = pva.angle;
+            auto robotCommand = robot->getNumtreePosControl()->getPosVelAngle(robot, ball->pos);
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/ballPlacement/BallPlacementPass.cpp]: Don't like unnecessary change to auto here.

File: roboteam_ai/src/skills/Skill.h [hunk 14]
```diff
@@ -6,15 +6,14 @@
 #include <roboteam_utils/Angle.h>
 #include "ros/ros.h"
 #include "../io/IOManager.h"
-#include "../control/positionControllers/PosVelAngle.h"
+#include "roboteam_ai/src/control/positionControllers/RobotCommand.h"
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/Skill.h]: Or, just include it in the relevant files? this smells like unnecessary compilation time since you're not changing the .cpp

File: roboteam_ai/src/skills/ShootFreeKick.cpp [hunk 15]
```diff
@@ -75,7 +75,8 @@ Skill::Status ShootFreeKick::onUpdate() {
         case SHOOTING: {
             if (! isShot()) {
                 Vector2 target = rtt::ai::world::field->getPenaltyPoint(false);
-                robot->getShotController()->makeCommand(robot->getShotController()->getShotData(* robot, target, true), command);
+                auto shotData = robot->getShotController()->getShotData(* robot, target, true);
+                command = shotData.makeROSCommand();
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/ShootFreeKick.cpp]: So if I understand correctly you're removing the call to the shotcontroller object that makes the command, so we're not keeping track correctly of the right variables for accelerationLimiter/VelocityLimiter anymore, as that is not done in the shotcontroller itself.

File: roboteam_ai/src/skills/Pass.cpp [hunk 16]
```diff
@@ -32,10 +32,10 @@ Pass::Status Pass::onUpdate() {
     bool closeToBall = (robot->pos - ball->pos).length() < CLOSE_ENOUGH_TO_BALL;
 
     if(!closeToBall && !passInitialized) {
-        auto pva = robot->getNumtreePosControl()->getPosVelAngle(robot, ball->pos);
-        command.x_vel = pva.vel.x;
-        command.y_vel = pva.vel.y;
-        command.w = pva.angle;
+        auto robotCommand = robot->getNumtreePosControl()->getPosVelAngle(robot, ball->pos);
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/Pass.cpp]: I'd rather see explicit typing where possible.

Comment (rolfvdhulst) [reply roboteam_ai/src/skills/Pass.cpp]: unless the type name is unnecessarily long which it really is not here.

File: roboteam_ai/src/skills/Attack.cpp [hunk 17]
```diff
@@ -22,16 +22,18 @@ Attack::Attack(string name, bt::Blackboard::Ptr blackboard)
...
         publishRobotCommand();
         return Status::Running;
     }
 
-    Vector2 aimPoint= coach::g_offensiveCoach.getShootAtGoalPoint(ball->pos);
-    robot->getShotController()->makeCommand(robot->getShotController()->getShotData(*robot, aimPoint,false,control::BallSpeed::MAX_SPEED,true,control::ShotPrecision::HIGH), command);
+    Vector2 aimPoint = coach::g_offensiveCoach.getShootAtGoalPoint(ball->pos);
+    auto shotData = robot->getShotController()->getShotData(
+            *robot, aimPoint, false, control::BallSpeed::MAX_SPEED, true, control::ShotPrecision::HIGH);
+    command = shotData.makeROSCommand();
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/Attack.cpp]: again this is not calling the velocity/acceleration limiter properly, I think.

File: roboteam_ai/src/control/shotControllers/ShotController.cpp [hunk 18]
```diff
@@ -265,39 +248,39 @@ int ShotController::determineOptimalGenevaState(world::Robot robot, Vector2 shot
...
-    shotData.forced = true;
+    shotData.kickerForced = true;
     return shotData;
 
 }
-ShotData ShotController::moveAndShoot(rtt::ai::world::Robot robot, bool chip,
+RobotCommand ShotController::moveAndShoot(rtt::ai::world::Robot robot, bool chip,
         std::pair<Vector2, Vector2> lineToDriveOver, BallSpeed desiredBallSpeed) {
-    control::PosVelAngle pva = robot.getBasicPosControl()->getPosVelAngle(std::make_shared<world::Robot>(robot),
+    auto robotCommand = robot.getBasicPosControl()->getPosVelAngle(std::make_shared<world::Robot>(robot),
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/shotControllers/ShotController.cpp]: make this a posVelAngle Type.

File: roboteam_ai/src/skills/Pass.cpp (hunk 16 above)
Comment (mrlukasbos) [reply roboteam_ai/src/skills/Pass.cpp]: Auto is fine

File: roboteam_ai/src/skills/Skill.h (hunk 14 above)
Comment (mrlukasbos) [reply roboteam_ai/src/skills/Skill.h]: Agreed with rolf

File: roboteam_ai/src/world/Robot.cpp [hunk 19]
```diff
@@ -192,6 +192,10 @@ void Robot::setPidPreviousVel(const Vector2 &pidVel) {
     pidPreviousVel = pidVel;
 }
 
+const Robot::RobotPtr Robot::copy() const {
```
[the remark this answers is not in the document]
Comment (mrlukasbos) [reply roboteam_ai/src/world/Robot.cpp]: Just use copy Constructors

File: roboteam_ai/src/world/Robot.h [hunk 20]
```diff
@@ -23,12 +23,14 @@ class BasicPosControl;
 
 namespace world {
 class Ball;
-class Robot {
+class Robot : public std::enable_shared_from_this<Robot> {
```
[the remark this answers is not in the document]
Comment (mrlukasbos) [reply roboteam_ai/src/world/Robot.h]: Agreed with Rolf I don't see any reason why you need this

File: roboteam_ai/src/skills/Pass.cpp (hunk 16 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/skills/Pass.cpp]: Yes it is not a problem, but why change the type from a known type to auto? I think it's fine if it is hard to read or a long declaration or so but should not be necessary here as it is just PosVelAngle

Comment (rolfvdhulst) [reply roboteam_ai/src/skills/Pass.cpp]: I think it's more readable if you can explicitly see the type that is returned in this case.

**Comment by mrlukasbos** [comment]
it is a bunch of magic but we need this
